avr-hal-generic/src/spi.rs

```diff
@@ -282,12 +318,301 @@ where
+        SPI,
+        port::Pin<port::mode::Output, SCLKPIN>,
+        port::Pin<port::mode::Output, MOSIPIN>,
+        port::Pin<port::mode::Input, MISOPIN>,
+    ) {
+        self.p.raw_release();
+        (self.p, self.sclk, self.mosi, self.miso)
+    }
+}
+
+impl<H, SPI, SCLKPIN, MOSIPIN, MISOPIN> embedded_hal::spi::ErrorType
+    for SpiBus<H, SPI, SCLKPIN, MOSIPIN, MISOPIN>
+where
+    SPI: SpiOps<H, SCLKPIN, MOSIPIN, MISOPIN>,
+    SCLKPIN: port::PinOps,
+    MOSIPIN: port::PinOps,
+    MISOPIN: port::PinOps,
+{
+    type Error = core::convert::Infallible;
+}
+
+impl<H, SPI, SCLKPIN, MOSIPIN, MISOPIN> embedded_hal::spi::SpiBus
+    for SpiBus<H, SPI, SCLKPIN, MOSIPIN, MISOPIN>
+where
+    SPI: SpiOps<H, SCLKPIN, MOSIPIN, MISOPIN>,
+    SCLKPIN: port::PinOps,
+    MOSIPIN: port::PinOps,
+    MISOPIN: port::PinOps,
+{
+    fn read(&mut self, read: &mut [u8]) -> Result<(), Self::Error> {
+        Ok(self.p.raw_read_buf(read))
+    }
+
+    fn write(&mut self, write: &[u8]) -> Result<(), Self::Error> {
+        Ok(self.p.raw_write_buf(write))
+    }
+
+    fn transfer(&mut self, read: &mut [u8], write: &[u8]) -> Result<(), Self::Error> {
+        match read.len().cmp(&write.len()) {
+            Ordering::Less => {
+                let (write1, write2) = write.split_at(read.len());
+                self.p.raw_read_write_buf(read, write1);
+                self.p.raw_write_buf(write2);
+            }
+            Ordering::Equal => self.p.raw_read_write_buf(read, write),
+            Ordering::Greater => {
+                let (read1, read2) = read.split_at_mut(write.len());
+                self.p.raw_read_write_buf(read1, write);
+                self.p.raw_read_buf(read2);
+            }
+        }
+        Ok(())
+    }
+
+    fn transfer_in_place(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        let write = unsafe {
+            let p = buffer.as_ptr();
+            core::slice::from_raw_parts(p, buffer.len())
+        };
+        self.p.raw_read_write_buf(buffer, write);
+        Ok(())
+    }
+
+    fn flush(&mut self) -> Result<(), Self::Error> {
+        if self.write_in_progress {
+            while !self.p.raw_check_iflag() {}
+            self.write_in_progress = false;
+        }
+        Ok(())
+    }
+}
+
+/// Behavior for a SPI interface in device mode (when using embedded-hal 1.0).
+///
+/// Stores the SPI peripheral for register access. In addition, it takes
+/// ownership of the MOSI and MISO pins to ensure they are in the correct mode.
+/// It also takes ownership of the CS pin to assert it when needed.
+/// Instantiate with the `new` method.
+pub struct SpiDevice<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN, CLOCK> {
+    p: SPI,
+    sclk: port::Pin<port::mode::Output, SCLKPIN>,
+    mosi: port::Pin<port::mode::Output, MOSIPIN>,
+    miso: port::Pin<port::mode::Input, MISOPIN>,
+    write_in_progress: bool,
+    cs: ChipSelectPin<CSPIN>,
+    delay: delay::Delay<CLOCK>,
+    _h: PhantomData<H>,
+}
```

Comment:
`SpiDevice` should only be implemented by a HAL when you can make use of hardware CS handling (See https://docs.rs/embedded-hal/latest/embedded_hal/spi/index.html#for-hal-authors).  As you are just taking an arbitrary pin here and calling `set_high()`, `set_low()`, there is no advantage of providing this `SpiDevice` implementation over the `ExclusiveDevice` from [`embedded-hal-bus`](https://docs.rs/embedded-hal-bus/latest/embedded_hal_bus/spi/index.html).